File: UpdaterZentrale-macOS/UpdaterZentrale/ViewModels/HauptViewModel.swift
import AppKit

/// Fensterlogik. 1:1-Port von ViewModels/HauptViewModel.cs.
///
/// Nicht portiert (kein macOS-Gegenstueck, siehe PORTING.md): `ImmerAlsAdmin`,
/// `AlsAdminNeuStarten`, `IstErhoeht`, `RechteText` -- macOS kennt keine Rechte-Erhoehung fuer
/// Programme, und Homebrew verweigert den Betrieb als root sogar ausdruecklich.
@MainActor
final class HauptViewModel {

    private let einstellungen = Einstellungen.laden()

    /// Alle bekannten Mechanismen, ueber das Katalogfeld "art" erreichbar. Ein neues Programm ist
    /// ein JSON-Eintrag; nur ein wirklich neuer Mechanismus braucht hier einen Eintrag.
    private let aktualisierer: [String: Aktualisierer] = [
        "brew": BrewAktualisierer(),
        "cli": CliAktualisierer(),
        "reposkript": RepoSkriptAktualisierer()
    ]

    /// Wird nach jeder Aenderung gefeuert, die das Fenster neu zeichnen muss.
    var beiAenderung: (() -> Void)?
    /// Wird gefeuert, wenn die Kartenliste ausgetauscht wurde (Katalog neu geladen).
    var beiListenWechsel: (() -> Void)?

    private(set) var programme: [ProgrammViewModel] = []

    var ausgewaehltesProgramm: ProgrammViewModel? {
        didSet { beiAenderung?() }
    }

    private(set) var kopfStatus = "Bereit." { didSet { beiAenderung?() } }
    private(set) var laeuftSammelvorgang = false { didSet { beiAenderung?() } }
    private(set) var katalogFehler: String? { didSet { beiAenderung?() } }

    init() {
        hellModusIntern = einstellungen.hellModus
        Darstellung.anwenden(hellModusIntern)
        katalogLaden()
    }

    // MARK: - Gruppierung

    /// Die Karten in der Reihenfolge des Katalogs, nach Gruppe gebuendelt.
    /// Ersetzt die CollectionViewSource-Gruppierung aus WPF.
    var gruppen: [(name: String, programme: [ProgrammViewModel])] {
        var reihenfolge: [String] = []
        var inhalt: [String: [ProgrammViewModel]] = [:]
        for programm in programme {
            if inhalt[programm.gruppe] == nil {
                inhalt[programm.gruppe] = []
                reihenfolge.append(programm.gruppe)
            }
            inhalt[programm.gruppe]?.append(programm)
        }
        return reihenfolge.map { ($0, inhalt[$0] ?? []) }
    }

    // MARK: - Darstellung

    private var hellModusIntern: Bool

    /// Hell oder dunkel; die Wahl wird in settings.json gemerkt.
    var hellModus: Bool {
        get { hellModusIntern }
        set {
            guard hellModusIntern != newValue else { return }
            hellModusIntern = newValue

            Darstellung.anwenden(newValue)
            einstellungen.hellModus = newValue
            einstellungen.speichern()

            for programm in programme { programm.darstellungAuffrischen() }
            beiAenderung?()
        }
    }

    var darstellungsText: String { hellModus ? "Heller Modus" : "Dunkler Modus" }

    // MARK: - Terminal

    private(set) var terminalAusgabe = "" { didSet { beiAenderung?() } }
    private(set) var terminalLaeuft = false { didSet { beiAenderung?() } }

    /// Kurz gehalten: daneben stehen noch zwei Schaltflaechen, und die rechte Spalte ist schmal.
    var terminalKopf: String { "Terminal · zsh" }

    func befehlAusfuehren(_ eingabe: String) async {
        let befehl = eingabe.trimmed
        guard !befehl.isEmpty, !terminalLaeuft else { return }

        terminalLaeuft = true
        terminalAusgabe += (terminalAusgabe.isEmpty ? "" : "\n\n") + "$ " + befehl + "\n"

        let ausgabe = await Terminal.ausfuehren(befehl, arbeitsverzeichnis: Terminal.arbeitsverzeichnis)
        terminalAusgabe += ausgabe
        terminalLaeuft = false
    }

    func terminalLeeren() { terminalAusgabe = "" }

    func terminalFensterOeffnen() {
        if !Terminal.fensterOeffnen() {
            Dialoge.hinweis("Es ließ sich kein Terminalfenster öffnen.")
        }
    }

    // MARK: - Kopf- und Fusszeile

    var anwendungsVersion: String {
        let info = Bundle.main.infoDictionary
        let version = (info?["CFBundleShortVersionString"] as? String) ?? "?"
        let stempel = info?["BuildTimestamp"] as? String
        let text = "Version " + version
        return stempel == nil ? text : text + "  ·  Build " + stempel!
    }

    var anzahlUpdates: Int { programme.filter(\.hatUpdate).count }

    var updateZusammenfassung: String {
        switch anzahlUpdates {
        case 0: return "Keine Updates offen"
        case 1: return "1 Update verfügbar"
        default: return "\(anzahlUpdates) Updates verfügbar"
        }
    }

    // MARK: - Katalog

    private func katalogLaden() {
        programme.removeAll()

        let (katalog, fehler) = Katalogdienst.laden()
        katalogFehler = fehler
        let berichte = Protokollierung.letzteBerichte()

        for eintrag in katalog.programme {
            let dienst = aktualisierer[eintrag.art.lowercased()]
            let karte = ProgrammViewModel(eintrag: eintrag, aktualisierer: dienst, einstellungen: einstellungen)
            karte.beiAenderung = { [weak self] in self?.beiAenderung?() }
            karte.beiMeldung = { text in Dialoge.hinweis(text) }

            // Der Verlauf ueberlebt Neustarts -- jede Karte kann zeigen, wie ihr letzter Lauf ging.
            if let bericht = berichte[eintrag.id] { karte.letzterBericht = bericht }

            programme.append(karte)
        }

        ausgewaehltesProgramm = programme.first
        beiListenWechsel?()
        beiAenderung?()
    }

    func katalogNeuLaden() {
        katalogLaden()
        kopfStatus = "Katalog neu geladen – \(programme.count) Programme."
    }

    /// Oeffnet programs.json, damit ein neues Programm ohne Eingriff in die App dazukommen kann.
    func katalogOeffnen() {
        let datei = Pfade.katalogDatei
        guard FileManager.default.fileExists(atPath: datei) else {
            Dialoge.hinweis("Der Katalog wurde nicht gefunden: \(datei)")
            return
        }
        NSWorkspace.shared.open(URL(fileURLWithPath: datei))
    }

    /// Oeffnet den Protokollordner; dort steht jeder Lauf, Tag fuer Tag.
    func protokolleOeffnen() { Protokollierung.ordnerOeffnen() }

    func zustaendeAuffrischen() {
        Systemdienst.zwischenspeicherLeeren()
        for programm in programme { programm.zustandAktualisieren() }
        kopfStatus = "Laufende Programme und Autostart wurden neu eingelesen."
    }

    // MARK: - Sammelvorgaenge

    /// Laeuft einmal, nachdem das Fenster steht -- so ist die Liste ohne ersten Klick aussagekraeftig,
    /// und die Update-Schaltflaechen zeigen ueberall dort "Aktuell", wo nichts ansteht.
    func erstePruefung() async {
        try? await Task.sleep(nanoseconds: 400_000_000)

        // Ist ein zuvor abgelegtes Update inzwischen angekommen -- oder haengt es noch?
        for programm in programme { await programm.ausstehendesPruefen() }

        await allePruefen()
    }

    func allePruefen() async {
        guard !laeuftSammelvorgang else { return }
        laeuftSammelvorgang = true
        defer { laeuftSammelvorgang = false }

        // Erst den Homebrew-Katalog auffrischen, DANN pruefen. Ohne das liest `brew info` den
        // lokalen Tap, und der kann Tage alt sein -- die Pruefung meldete dann "Aktuell", obwohl
        // ein Cask laengst neuer ist. `winget list` unter Windows hat dieses Problem nicht, weil es
        // seine Quelle selbst aktuell haelt.
        if programme.contains(where: { $0.eintrag.art == "brew" }) {
            kopfStatus = "Homebrew-Katalog wird aufgefrischt …"
            _ = await Kommandozeile.ausfuehren(Pfade.brew, ["update", "--quiet"], zeitlimit: 300)
        }

        // Bewusst nacheinander: brew greift ohnehin seriell auf seinen Katalog zu, und ein
        // paralleler Schwung macht das Protokoll unlesbar.
        let gesamt = programme.count
        for (i, programm) in programme.enumerated() {
            kopfStatus = "Prüft \(programm.name) (\(i + 1) von \(gesamt)) …"
            await programm.pruefen()
        }
        kopfStatus = "Prüfung abgeschlossen – " + updateZusammenfassung + "."
    }

    func alleAktualisieren() async {
        guard !laeuftSammelvorgang else { return }

        let offen = programme.filter(\.hatUpdate)
        guard !offen.isEmpty else {
            Dialoge.hinweis("Es ist kein Update offen. Prüfe zuerst, oder aktualisiere einzelne Programme gezielt.")
            return
        }

        let liste = offen.map { "  • " + $0.name }.joined(separator: "\n")
        guard Dialoge.fragen("Diese Programme werden jetzt aktualisiert:\n\n\(liste)"
                             + "\n\nBei laufenden Programmen wird vorher nachgefragt.",
                             "Alle Updates installieren?") else { return }

        laeuftSammelvorgang = true
        defer { laeuftSammelvorgang = false }

        for (i, programm) in offen.enumerated() {
            kopfStatus = "Aktualisiert \(programm.name) (\(i + 1) von \(offen.count)) …"
            ausgewaehltesProgramm = programm
            await programm.aktualisieren()
        }
        kopfStatus = "Alle Updates sind durchgelaufen."
    }
}

File: UpdateZentrale/ViewModels/HauptViewModel.cs
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Diagnostics;
using System.Reflection;
using System.Windows.Data;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using UpdateZentrale.Models;
using UpdateZentrale.Providers;
using UpdateZentrale.Services;

namespace UpdateZentrale.ViewModels;

public sealed partial class HauptViewModel : ObservableObject
{
    private readonly Einstellungen _einstellungen = Einstellungen.Laden();

    /// <summary>
    /// All known mechanisms, keyed by the catalog field "art". A new program is a JSON entry;
    /// only a genuinely new mechanism needs a new entry here.
    /// </summary>
    private readonly Dictionary<string, IAktualisierer> _aktualisierer =
        new(StringComparer.OrdinalIgnoreCase)
        {
            ["winget"] = new WingetAktualisierer(),
            ["cli"] = new CliAktualisierer(),
            ["store"] = new StoreAktualisierer(),
            ["msstore"] = new StoreAktualisierer(),   // Altname aus früheren Katalogfassungen
            ["reposkript"] = new RepoSkriptAktualisierer()
        };

    public HauptViewModel()
    {
        Programme = new ObservableCollection<ProgrammViewModel>();
        Ansicht = CollectionViewSource.GetDefaultView(Programme);
        Ansicht.GroupDescriptions.Add(new PropertyGroupDescription(nameof(ProgrammViewModel.Gruppe)));

        _hellModus = _einstellungen.HellModus;
        Darstellung.Anwenden(_hellModus);

        // The updater is meant to be able to install anything, so it marks itself to always run
        // elevated on the first launch. The switch in the footer turns that off again.
        if (!_einstellungen.AdminStartGesetzt)
        {
            Rechte.ImmerAlsAdminSetzen(true);
            _einstellungen.AdminStartGesetzt = true;
            _einstellungen.Speichern();
        }
        _immerAlsAdmin = Rechte.ImmerAlsAdmin;

        KatalogLaden();
    }

    public ObservableCollection<ProgrammViewModel> Programme { get; }

    public ICollectionView Ansicht { get; }

    [ObservableProperty] private ProgrammViewModel? _ausgewaehltesProgramm;
    [ObservableProperty] private string _kopfStatus = "Bereit.";
    [ObservableProperty] private bool _laeuftSammelvorgang;
    [ObservableProperty] private string? _katalogFehler;

    private bool _hellModus;

    /// <summary>Light or dark; the choice is remembered in settings.json.</summary>
    public bool HellModus
    {
        get => _hellModus;
        set
        {
            if (_hellModus == value) return;
            _hellModus = value;

            Darstellung.Anwenden(value);
            _einstellungen.HellModus = value;
            _einstellungen.Speichern();

            // The state pills are drawn by converters that read the mode, so they need a nudge.
            foreach (var p in Programme) p.DarstellungAuffrischen();

            OnPropertyChanged();
            OnPropertyChanged(nameof(DarstellungsText));
        }
    }

    public string DarstellungsText => HellModus ? "Heller Modus" : "Dunkler Modus";

    // ---------------- Rechte ----------------

    private bool _immerAlsAdmin;

    public bool IstErhoeht => Rechte.IstErhoeht;
    public bool NichtErhoeht => !Rechte.IstErhoeht;

    public string RechteText => Rechte.IstErhoeht
        ? "Läuft mit Administratorrechten"
        : "Läuft ohne Administratorrechte";

    /// <summary>Marks the exe itself with the RUNASADMIN compatibility flag.</summary>
    public bool ImmerAlsAdmin
    {
        get => _immerAlsAdmin;
        set
        {
            if (_immerAlsAdmin == value) return;
            _immerAlsAdmin = value;

            if (!Rechte.ImmerAlsAdminSetzen(value))
            {
                _immerAlsAdmin = !value;
                Dialoge.Hinweis("Die Einstellung ließ sich nicht schreiben.");
            }
            OnPropertyChanged();
        }
    }

    [RelayCommand]
    private void AlsAdminNeuStarten()
    {
        if (Rechte.IstErhoeht)
        {
            Dialoge.Hinweis("Die UpdateZentrale läuft bereits mit Administratorrechten.");
            return;
        }

        if (Rechte.NeuStartenAlsAdmin())
        {
            System.Windows.Application.Current?.Shutdown();
        }
        else
        {
            KopfStatus = "Der Neustart mit Administratorrechten wurde abgebrochen.";
        }
    }

    // ---------------- Terminal ----------------

    [ObservableProperty] private string _befehl = "";
    [ObservableProperty] private string _terminalAusgabe = "";
    [ObservableProperty] private bool _terminalLaeuft;

    public string TerminalKopf => "Terminal · " + (Rechte.IstErhoeht ? "Administrator" : "Standardrechte");

    [RelayCommand]
    private async Task BefehlAusfuehrenAsync()
    {
        var befehl = Befehl.Trim();
        if (string.IsNullOrWhiteSpace(befehl) || TerminalLaeuft) return;

        TerminalLaeuft = true;
        TerminalAusgabe += (TerminalAusgabe.Length == 0 ? "" : "\n\n") + "PS> " + befehl + "\n";
        Befehl = "";

        try
        {
            var ausgabe = await Terminal.AusfuehrenAsync(befehl);
            TerminalAusgabe += ausgabe;
        }
        catch (Exception ex)
        {
            TerminalAusgabe += "[Fehler] " + ex.Message;
        }
        finally
        {
            TerminalLaeuft = false;
        }
    }

    [RelayCommand]
    private void TerminalLeeren() => TerminalAusgabe = "";

    [RelayCommand]
    private void TerminalFensterOeffnen()
    {
        if (!Terminal.FensterOeffnen())
        {
            Dialoge.Hinweis("Es ließ sich kein Terminalfenster öffnen.");
        }
    }

    public string AnwendungsVersion
    {
        get
        {
            var version = Assembly.GetExecutingAssembly().GetName().Version;
            var stempel = Assembly.GetExecutingAssembly()
                .GetCustomAttributes<AssemblyMetadataAttribute>()
                .FirstOrDefault(a => a.Key == "BuildTimestamp")?.Value;

            var text = "Version " + (version is null ? "?" : version.ToString(3));
            return stempel is null ? text : text + "  ·  Build " + stempel;
        }
    }

    public int AnzahlUpdates => Programme.Count(p => p.HatUpdate);

    public string UpdateZusammenfassung => AnzahlUpdates switch
    {
        0 => "Keine Updates offen",
        1 => "1 Update verfügbar",
        _ => AnzahlUpdates + " Updates verfügbar"
    };

    private void KatalogLaden()
    {
        foreach (var alt in Programme) alt.PropertyChanged -= AufProgrammGeaendert;
        Programme.Clear();

        var (katalog, fehler) = Katalogdienst.Laden();
        KatalogFehler = fehler;
        var berichte = Protokollierung.LetzteBerichte();

        foreach (var eintrag in katalog.Programme)
        {
            _aktualisierer.TryGetValue(eintrag.Art, out var dienst);
            var vm = new ProgrammViewModel(eintrag, dienst, _einstellungen);
            vm.PropertyChanged += AufProgrammGeaendert;
            vm.Meldung += (_, text) => Dialoge.Hinweis(text);

            // The history survives restarts, so each card can show how its last run went.
            if (berichte.TryGetValue(eintrag.Id, out var bericht)) vm.LetzterBericht = bericht;

            Programme.Add(vm);
        }

        AusgewaehltesProgramm = Programme.FirstOrDefault();
        OnPropertyChanged(nameof(AnzahlUpdates));
        OnPropertyChanged(nameof(UpdateZusammenfassung));
    }

    private void AufProgrammGeaendert(object? sender, PropertyChangedEventArgs e)
    {
        if (e.PropertyName is nameof(ProgrammViewModel.Zustand) or nameof(ProgrammViewModel.HatUpdate))
        {
            OnPropertyChanged(nameof(AnzahlUpdates));
            OnPropertyChanged(nameof(UpdateZusammenfassung));
        }
    }

    /// <summary>
    /// Runs once after the window is up, so the list is meaningful without a first click -- and
    /// so the update buttons show "Aktuell" wherever nothing is pending.
    /// </summary>
    public async Task ErstePruefungAsync()
    {
        await Task.Delay(400);

        // Which programs already start elevated through a scheduled task?
        foreach (var p in Programme) await p.AufgabenZustandLesenAsync();

        // Did a previously staged update arrive in the meantime -- or is it still hanging?
        foreach (var p in Programme) await p.AusstehendesPruefenAsync();

        await AllePruefenAsync();
    }

    [RelayCommand]
    private async Task AllePruefenAsync()
    {
        if (LaeuftSammelvorgang) return;
        LaeuftSammelvorgang = true;
        try
        {
            // Sequential on purpose: winget serialises its source access anyway, and a parallel
            // burst makes the log unreadable.
            // A snapshot: the list must not shift under the loop (reload is blocked meanwhile,
            // this is the second layer).
            var liste = Programme.ToList();
            var gesamt = liste.Count;
            for (var i = 0; i < gesamt; i++)
            {
                var p = liste[i];
                KopfStatus = "Prüft " + p.Name + " (" + (i + 1) + " von " + gesamt + ") …";
                await p.PruefenCommand.ExecuteAsync(null);
            }
            KopfStatus = "Prüfung abgeschlossen – " + UpdateZusammenfassung + ".";
        }
        finally
        {
            LaeuftSammelvorgang = false;
        }
    }

    [RelayCommand]
    private async Task AlleAktualisierenAsync()
    {
        if (LaeuftSammelvorgang) return;

        var offen = Programme.Where(p => p.HatUpdate).ToList();
        if (offen.Count == 0)
        {
            Dialoge.Hinweis("Es ist kein Update offen. Prüfe zuerst, oder aktualisiere einzelne Programme gezielt.");
            return;
        }

        var liste = string.Join("\n", offen.Select(p => "  • " + p.Name));
        if (!Dialoge.Fragen("Diese Programme werden jetzt aktualisiert:\n\n" + liste
                            + "\n\nBei laufenden Programmen wird vorher nachgefragt.", "Alle Updates installieren?"))
        {
            return;
        }

        LaeuftSammelvorgang = true;
        try
        {
            for (var i = 0; i < offen.Count; i++)
            {
                KopfStatus = "Aktualisiert " + offen[i].Name + " (" + (i + 1) + " von " + offen.Count + ") …";
                AusgewaehltesProgramm = offen[i];
                await offen[i].AktualisierenCommand.ExecuteAsync(null);
            }
            KopfStatus = "Alle Updates sind durchgelaufen.";
        }
        finally
        {
            LaeuftSammelvorgang = false;
        }
    }

    [RelayCommand]
    private void KatalogNeuLaden()
    {
        // Reloading throws the cards away. A card whose update is still running would vanish
        // mid-run -- its result never shown, and a second run of the same installer possible
        // from the fresh card.
        if (LaeuftSammelvorgang || Programme.Any(p => p.IstBeschaeftigt))
        {
            KopfStatus = "Neu laden geht erst, wenn alle laufenden Prüfungen und Updates fertig sind.";
            return;
        }

        KatalogLaden();
        KopfStatus = "Katalog neu geladen – " + Programme.Count + " Programme.";
    }

    /// <summary>Opens programs.json so a new program can be added without touching the app.</summary>
    [RelayCommand]
    private void KatalogOeffnen()
    {
        try
        {
            Process.Start(new ProcessStartInfo { FileName = Pfade.KatalogDatei, UseShellExecute = true });
        }
        catch (Exception ex)
        {
            Dialoge.Hinweis("Der Katalog ließ sich nicht öffnen: " + ex.Message);
        }
    }

    /// <summary>Opens the log folder; every run is recorded there, day by day.</summary>
    [RelayCommand]
    private void ProtokolleOeffnen() => Protokollierung.OrdnerOeffnen();

    [RelayCommand]
    private void ZustaendeAuffrischen()
    {
        foreach (var p in Programme) p.ZustandAktualisieren();
        KopfStatus = "Laufende Programme und Autostart wurden neu eingelesen.";
    }
}
